alot/db/manager.py

```diff
@@ -225,7 +257,7 @@ def get_thread(self, tid):
         return Thread(self, self._get_notmuch_thread(tid))
 
     def _get_notmuch_message(self, mid):
-        """returns :class:`notmuch.database.Message` with given id"""
+        """returns :class:`notmuch2.Message` with given id"""
```

Comment:
Is this function an example of where you have a problem when you use the context manager of notmuch2.Database?

As far as I can see the old bindings call `notmuch_database_destroy()` on `__del__()`.  This means as soon as this function returns the database could be destroyed, where the cffi bindings only call `notmuch_database_close()` when you use a context manager.  So I don't really understand yet why your message object would behave much worse, if anything I would expect it to behave better.  And as far as I understand under CPython I would expect `db.__del__()` to be called as soon as this function returns.
The only way I can see this work better is if for some reason I don't understand yet the `db.__del__()` call was delayed for some reason.

Reply:
No, the issue I noticed stems from there:

https://github.com/pazz/alot/blob/34929e49ccfde98209ffa22106df32a5ce11ee14/alot/db/manager.py#L246-L253

When the database connection is done with a context manager, it seems that the iterator returned in this function becomes invalid (points to zeroed-out data, for example the dates are the epoch, the strings are empty etc.).

Reply:
Which version of python does the filedescriptor leakage happen on?  And is it possible to see if there is anything in `gc.garbage` when you have the leaks?

Reply:
I'm using Python `v3.8.3`.

I've identified the source of the leak to be successive calls (two) to the function above, `_get_notmuch_thread()`.

I simply counted the file descriptors open within the current process on several lines, and proceeding by dichotomy I found out that a thread object and its sub-objects call that function, which at some point leaks.

The `gc.garbage` list is empty after the leak occurred, and the file descriptors are never closed even when a thread is deleted.

Reply:
I just can't get a reproducer to highlight the issue, even though I clearly see it in Alot.

Reply:
any news on this particular memory leak issue?

Reply:
Last time I looked at this I didn't gain any insights to what was happening.  I'll try and have a look again in the next week and try to understand alot's design a bit more to see if I can come up with some better understanding.

Reply:
Does this help: https://github.com/lenormf/alot/pull/1 ?